src/main/scala/lms/thirdparty/cuda/cuda.scala

```diff
@@ -1175,6 +1178,113 @@ trait CudaLibs extends CudaOps {
+    val input_size = dimZ * dimY * dimX
+    val offsets = dimXs.scanLeft(0) { case (acc, x) => acc + x }
+    val n = dimXs.length
+
+    (in: Rep[Array[N]], out: Rep[Array[Array[N]]]) => {
+      generate_comment(s"This is cuda $n-section split kernel for 3D input at axis 2.")
+      generate_comment(s"It takes a 3D array and splits on the innermost dimension (dim2) into $n arrays.")
+      generate_comment("arg0: input array")
+      generate_comment("arg1: array of output arrays")
+      generate_comment(s"call constraint: sum of out(i).size = in.size for i in [0, $n)")
+
+      val idx = blockIdxX * blockDimX + threadIdxX
+
+      __ifThenElse(idx < input_size, {
+        val value = in(idx)
+        val x = idx / dimX
+        val y = idx % dimX
+
+        def get_case(t: Int) = {
+          val temp = out(t)
+          cudaArrayEffect[N](temp, x * dimXs(t) + (y - offsets(t)), value)
+        }
+
+        def make_case(t: Int): Unit = {
+          if (t == n-1) {
+            get_case(t)
+          } else {
+            __ifThenElse(y < offsets(t+1), { get_case(t) }, { make_case(t+1) })
+          }
+        }
+        make_case(0)
+      }, {})
+    }
+  }
+
+  def cuda3DSplitWrap[N:Numeric:Manifest](in: Rep[Array[N]], outs: List[Rep[Array[N]]], dimZ: Int, dimY: Int, dimXs: List[Int], grid: Rep[Dim3], block: Rep[Dim3])(implicit __pos: SourceContext) = {
+    val sec = outs.length
+    val output = NewArray[Array[N]](sec)
+    for (i <- (0 until sec): Range) {
+      output(i) = outs(i)
+    }
+    val cuda_output = cudaMalloc2[Array[N]](sec)
+    cudaCall(cudaMemcpyOfT[Array[N]](cuda_output, output, sec, host2device))
+
+    val splitKernel = cuda3DSplitAxis2[N](dimZ, dimY, dimXs)
+    splitKernel(in, cuda_output, grid, block)
+  }
+
+  // This is the kernel function for concatenating 3D inputs at the inner most axis
+  // the input i is of shape (dimZ, dimY, dimXs(i))
+  // the output shape is (dimZ, dimY, sum(dimXs))
+  def cuda3DConcatAxis2[N:Numeric:Manifest](dimZ: Int, dimY: Int, dimXs: List[Int])(implicit __pos: SourceContext) = cudaGlobalFun {
+    val dimX = dimXs.reduce(_ + _)
+    val d_other = dimZ * dimY
+    val input_size = d_other * dimX
+    val offsets = dimXs.scanLeft(0) { case (acc, x) => acc + x }
+    val n = dimXs.length
+
+    (in: Rep[Array[Array[N]]], out: Rep[Array[N]]) => {
+      generate_comment(s"this is cuda $n-section concat kernel for 3D inputs at axis 2.")
+      generate_comment(s"It concatenates $n 3D arrays on the innermost dimension (dim2).")
+      generate_comment("arg0: array of input input arrays")
+      generate_comment("arg1: output array")
+      generate_comment(s"call constraint: in.size = $n")
+      generate_comment(s"call constraint: sum of in(i).size = out.size for i in [0, $n)")
+
+      val idx = blockIdxX * blockDimX + threadIdxX
+
+      __ifThenElse(idx < input_size, {
+        val x = idx / dimX
+        val y = idx % dimX
+
+        def get_case(t: Int) = {
```

Comment:
maybe `get_case` and `make_case` can be changed to better names?

Reply:
I agree. let me fix them